### PR TITLE
Add filmtipset ratings parser

### PR DESCRIPTION
**Pull request by Row**
### Type of Change

new feature

### Issue or RFC
Add source www.filmtipset.se #97

### Description of the Change
Add filmtipset site and ratings parser, which use filmtipset's csv export functionality. 

### Possible Drawbacks
filmtipset's csv is malformed

### CODE REVIEW

**Comment by Row** [comment]
I am still participating in hacktoberfest this year so would be great if this PR counted towards the quota.

> Your PR was submitted to a repository that is not participating in Hacktoberfest. Maintainers of the repository can add the "hacktoberfest" topic to their repository if they wish to participate. Alternatively, an individual PR can be opted-in with a maintainer adding the "hacktoberfest-accepted" label to the PR.

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        self.site.browser.find_element_by_xpath('//input[@id="submit-export"]').click()
+        self.site.browser.get('https://www.filmtipset.se/export-data.php?q=')
+
+    #VoteDate;MovieTitle;IMDB;Score
+    def _convert_csv_row_to_movie(self, headers, row):
+        movie = dict()
+        movie['title'] = row[headers.index("MovieTitle")]
+        movie[self.site.site_name.lower()] = dict()
+        filmtipset_rating = int(row[headers.index("Score")])
+        movie[self.site.site_name.lower()]['my_rating'] = filmtipset_rating
```
Comment (Row) [inline RatS/filmtipset/filmtipset_ratings_parser.py:64]: I am not sure if I should keep this as the original 1 - 5 grade or not.

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            i = int(imdb_id)
+            if i < 1:
+                raise ValueError
+        except ValueError:
+            return
+
+        imdb = dict()
+        imdb['id'] = 'tt{imdb_id_number:07d}'.format(imdb_id_number=int(imdb_id))
+        imdb['url'] = 'https://www.imdb.com/title/{imdb_id}'.format(imdb_id=imdb['id'])
+        imdb['my_rating'] = int(filmtipset_rating * 2)
```
Comment (Row) [inline RatS/filmtipset/filmtipset_ratings_parser.py:81]: I transform the rating to 1 - 10 here and it's not clear to me if this is the right approach. But it feels more correct than adding the 1 - 10 rating in the filmtipset dict.

File: RatS/filmtipset/filmtipset_ratings_parser.py (hunk 1 above)
Comment (StegSchreck) [reply RatS/filmtipset/filmtipset_ratings_parser.py:64]: So far, I always translated the site's rating scale to a unified 1-10 scale. Only this way, we can be sure that a 3 in filmtipset is the same as a 3 in Trakt.

Comment (StegSchreck) [reply RatS/filmtipset/filmtipset_ratings_parser.py:64]:
```suggestion
        movie[self.site.site_name.lower()]['my_rating'] = filmtipset_rating * 2
```

File: RatS/filmtipset/filmtipset_ratings_parser.py (hunk 2 above)
Comment (StegSchreck) [reply RatS/filmtipset/filmtipset_ratings_parser.py:81]: Semantically, the `movie['imdb']['my_rating']` is meant to be the rating I gave to the move on the IMDB page. 
For a transfer from Filmtipset to e.g. Trakt, this value would be ignored and the `movie['filmtipset']['my_rating']` is the source value for rating the same movie on the destination page.

```suggestion
```

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        self.downloaded_file_name = 'ft_betyg_{}.csv'.format(date)
+        self.csv_delimiter = ';'
+
+    # TODO remove when filmtipset has cleaned up thier "csv" file
+    @staticmethod
+    def _repair_line(line):
+        output = re.search(r'([0-9]{4}-[0-9]{2}-[0-9]{2}),(.*)(;[\-0-9]*;[1-5])$', line)
+        if output is None:
+            return line
+        group2 = output.group(2).replace('"', '""')
```
Comment (StegSchreck) [inline RatS/filmtipset/filmtipset_ratings_parser.py:37]: can we give `group2` a more meaningful name?

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import os
+import datetime
+import time
+import fileinput
+import re
+from shutil import copyfile
+from RatS.base.base_ratings_downloader import RatingsDownloader
+from RatS.filmtipset.filmtipset_site import Filmtipset
+from RatS.utils import file_impex
```
Comment (StegSchreck) [inline RatS/filmtipset/filmtipset_ratings_parser.py:9]: - [ ] remove unused import

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+# Request download form
```
Comment (StegSchreck) [inline RatS/filmtipset/filmtipset_ratings_parser.py:11]: - [ ] remove this comment, as how the form looks like in the HTML is not relevant and quickly outdated

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        super(FilmtipsetRatingsParser, self).__init__(Filmtipset(args), args)
+        # TODO: due to timezone and potential racecondition, it's probably more safe to
+        #       find/search file in download folder
+        date = datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d')
+        self.downloaded_file_name = 'ft_betyg_{}.csv'.format(date)
+        self.csv_delimiter = ';'
+
+    # TODO remove when filmtipset has cleaned up thier "csv" file
+    @staticmethod
+    def _repair_line(line):
```
Comment (StegSchreck) [inline RatS/filmtipset/filmtipset_ratings_parser.py:33]: For the sake of consistency, I'd like to suggest this name
```suggestion
    def _repair_csv_line(line):
```

Actually, I used the term "row" for the data lines of the csv in other places and suggest to apply this name here too.

- [ ] consider renaming `line` to `row` when referencing the csv items

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        output = re.search(r'([0-9]{4}-[0-9]{2}-[0-9]{2}),(.*)(;[\-0-9]*;[1-5])$', line)
+        if output is None:
+            return line
+        group2 = output.group(2).replace('"', '""')
+        newline = '{};"{}"{}\n'.format(output.group(1), group2, output.group(3))
+        return newline
```
Comment (StegSchreck) [inline RatS/filmtipset/filmtipset_ratings_parser.py:39]:
```suggestion
        repaired_line = '{};"{}"{}\n'.format(output.group(1), group2, output.group(3))
        return repaired_line
```
(or `repaired_row`, see my other comment)

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+    @staticmethod
+    def _repair_csv_file(filepath):
+        copyfile(filepath, filepath + '.bak')
+        for line in fileinput.input(filepath, inplace=True): 
+            print(FilmtipsetRatingsParser._repair_line(line),end="")
```
Comment (StegSchreck) [inline RatS/filmtipset/filmtipset_ratings_parser.py:45]:
```suggestion
            print(FilmtipsetRatingsParser._repair_line(line), end="")
```

- [ ] Please check all the PEP8 warnings in this file

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        self._rename_csv_file(self.downloaded_file_name)
+        filepath = os.path.join(self.exports_folder, self.csv_filename)
+        self._repair_csv_file(filepath)
+        self.movies = self._parse_movies_from_csv(filepath)
+
+    def _call_download_url(self):
+        self.site.browser.find_element_by_xpath('//input[@id="submit-export"]').click()
+        self.site.browser.get('https://www.filmtipset.se/export-data.php?q=')
+
+    #VoteDate;MovieTitle;IMDB;Score
```
Comment (StegSchreck) [inline RatS/filmtipset/filmtipset_ratings_parser.py:58]:
```suggestion
```

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    def _convert_csv_row_to_movie(self, headers, row):
+        movie = dict()
+        movie['title'] = row[headers.index("MovieTitle")]
+        movie[self.site.site_name.lower()] = dict()
+        filmtipset_rating = int(row[headers.index("Score")])
```
Comment (StegSchreck) [inline RatS/filmtipset/filmtipset_ratings_parser.py:63]: - [ ] consider renaming `filmtipset_rating` -> `my_rating`

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    def _parse_ratings(self):
+        self._download_ratings_csv()
+        self._rename_csv_file(self.downloaded_file_name)
+        filepath = os.path.join(self.exports_folder, self.csv_filename)
+        self._repair_csv_file(filepath)
+        self.movies = self._parse_movies_from_csv(filepath)
+
+    def _call_download_url(self):
+        self.site.browser.find_element_by_xpath('//input[@id="submit-export"]').click()
+        self.site.browser.get('https://www.filmtipset.se/export-data.php?q=')
```
Comment (StegSchreck) [inline RatS/filmtipset/filmtipset_ratings_parser.py:56]: When I tested locally, I noticed that there was a file named `ft_betyg_2020-10-14(1).csv` left in the file system. And I think it is related to these two lines. One of them is obsolete.

File: tests/unit/filmtipset/test_filmtipset_ratings_parser.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+            ('1234567', {'imdb': {'id': 'tt1234567', 'url': 'https://www.imdb.com/title/tt1234567'}}),
+            ('12345', {'imdb': {'id': 'tt0012345', 'url': 'https://www.imdb.com/title/tt0012345'}}),
+            ('', {})
+        ]
+        for test in tests:
+            movie = dict()
+            num, expected_movie  = test
+            FilmtipsetRatingsParser._extract_imdb_informations(movie, num) # pylint: disable=protected-access
+            self.assertDictEqual(movie, expected_movie)
+
```
Comment (StegSchreck) [inline tests/unit/filmtipset/test_filmtipset_ratings_parser.py:100]: Remove spaces here, but leave blank line
```suggestion

```

File: tests/unit/filmtipset/test_filmtipset_ratings_parser.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    @patch('RatS.base.base_ratings_parser.RatingsParser.__init__')
+    @patch('RatS.filmtipset.filmtipset_ratings_parser.Filmtipset')
+    def test_convert_csv_row_to_movie(self, site_mock, base_init_mock, browser_mock, imdb_mock):
+        parser = FilmtipsetRatingsParser(None)
+        parser.movies = []
+        parser.site = site_mock
+        parser.site.site_name = 'Filmtipset'
+        parser.site.browser = browser_mock
+
+        expt_movie = {'title': 'Hunt for the Wilderpeople', 'filmtipset': {'my_rating': 10}}
```
Comment (StegSchreck) [inline tests/unit/filmtipset/test_filmtipset_ratings_parser.py:82]: please don't use this kind of abbreviations, as it makes the code less readable.

```suggestion
        expected_movie = {'title': 'Hunt for the Wilderpeople', 'filmtipset': {'my_rating': 10}}
```

**Review by StegSchreck — CHANGES_REQUESTED** [review]
Hey @Row 
Thank you very much for your contribution!
Please have a look at the comments I added.

Furthermore, I'd like you to consider:
- [ ] Make all files PEP8 compatible
- [ ] Adjust README file for the new parser (don't worry about the image in there, I will take care of that later)

File: RatS/filmtipset/filmtipset_ratings_parser.py (hunk 11 above)
Comment (Row) [reply RatS/filmtipset/filmtipset_ratings_parser.py:56]: Yes, you are right the file is downloaded with javascript and window.open, I was not sure if it was started or not.
Will remove this.

File: RatS/filmtipset/filmtipset_ratings_parser.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        self._extract_imdb_informations(movie, row[headers.index("IMDB")], filmtipset_rating)
+
+        return movie
+
+    @staticmethod
+    def _extract_imdb_informations(movie, imdb_id, filmtipset_rating):
+        try:
+            i = int(imdb_id)
+            if i < 1:
+                raise ValueError
```
[the remark this answers is not in the document]
Comment (Row) [reply RatS/filmtipset/filmtipset_ratings_parser.py:74]: Yes, true. But the line above may still raise ValueError.

File: tests/assets/filmtipset/ft_betyg_2020-10-10.csv [hunk 15]
```diff
@@ -0,0 +1,8 @@
+VoteDate;MovieTitle;IMDB;Score
```
[the remark this answers is not in the document]
Comment (Row) [reply tests/assets/filmtipset/ft_betyg_2020-10-10.csv:1]: The file is downloaded as ft_betyg_2020-10-10.csv, is it possible to change the name when it's downloaded? Should not the asset reflect what is fetched from the live site?
Or do you mean that the test should make a copy of my_ratings.csv to e.g. ft_betyg_2020-10-10.csv?

[the remark this answers is not in the document]
Comment (StegSchreck) [reply tests/assets/filmtipset/ft_betyg_2020-10-10.csv:1]: I've seen that Filmtipset's default name is like this. 
What I meant: in the test, the filename can be something more readable. Have a look at https://github.com/StegSchreck/RatS/blob/master/tests/unit/movielens/test_movielens_ratings_parser.py#L35 for reference.

[the remark this answers is not in the document]
Comment (Row) [reply tests/assets/filmtipset/ft_betyg_2020-10-10.csv:1]: I have changed the asset name to my_grades.csv. I have also implemented the file poll to search for the ft_betyg-yyyy-mm-dd. Will update the PR later to night or at latest during the weekend.

p.s. fyi "betyg" is "grade/rating" in Swedish :)

[the remark this answers is not in the document]
Comment (Row) [reply tests/assets/filmtipset/ft_betyg_2020-10-10.csv:1]: But I guess I can create any empty file with the name ft_betyg_2020-10-10.csv

**Comment by Row** [comment]
I am not sure I managed to get the PEP8 thing to work on OSX, got a notice that PEP8 was deprecated and had to use `pycodestyle` instead.
In the test it complained about long lines, I fix a lot of those but I later noticed that pytest.ini has options to PEP8 to ignore this? Had to use `pycodestyle . | grep filmtipset` to find my lint errors.  What this means is that there might be other errors that I can not see in my setup.  It looks like the CI log is ok.

**Pull request merged**